kani-compiler/src/kani_middle/stubbing/annotations.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//! This file contains code for extracting stubbing-related attributes.
+
+use rustc_ast::Attribute;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_driver::RunCompiler;
+use rustc_driver::{Callbacks, Compilation};
+use rustc_errors::ErrorGuaranteed;
+use rustc_interface::interface::Compiler;
+use rustc_interface::Queries;
+use rustc_middle::ty::TyCtxt;
+
+use crate::kani_middle::attributes::{extract_path_arguments, partition_kanitool_attributes};
+
+/// Collects the stubs from the harnesses in a crate, running rustc with the
+/// supplied arguments `rustc_args`.
+pub fn collect_stub_mappings(
+    rustc_args: &[String],
+) -> Result<FxHashMap<String, FxHashMap<String, String>>, ErrorGuaranteed> {
+    let mut callbacks = CollectorCallbacks::default();
+    let compiler = RunCompiler::new(rustc_args, &mut callbacks);
+    compiler.run().map(|_| callbacks.stub_mapping)
+}
+
+/// A rustc callback that is used to collect the stub mappings specified for
+/// each harness.
+#[derive(Default)]
+struct CollectorCallbacks {
+    stub_mapping: FxHashMap<String, FxHashMap<String, String>>,
+}
+
+impl Callbacks for CollectorCallbacks {
```

Comment:
Could you expand on this comment? The callback is used for the local call to rustc.

Reply:
I guess my concern is that this call is kind of hidden and it may catch developers by surprise that rustc is being invoked from this submodule.

Reply:
Fair. Some options:

1. Keep this preliminary call to rustc, but move the call itself into `main`, so that it is more visible.
2. Remove this call to rustc, and run the annotation collecting callback in the same rustc call that does code gen. I originally went with two calls to rustc because it'd allow us to figure out which harnesses use the same stub set, and then we'd able to code gen those ones together. However, if we are going to be doing per-harness code gen, figuring out the stub sets first doesn't really get us anything.

Reply:
I think it's OK for this PR.